### ISSUE TITLE
Initial automated deployment of containers

### ISSUE THREAD

**Issue by mikehaller** [eclipse-kanto/container-management#95]
I'd like to describe the initial set of containers to be deployed when container-management is starting up for the first time.

It would be great if container-management could read in container deployment descriptors from the filesystem to perform initial pulling of images and creation of containers on initial startup of a device.

Requirements:
- Define image and additional configuration (container name, image ref, mounts, ports, environment variables etc.) in a json deployment descriptor
- Configure the container manager to iterate through all deployment descriptors in a given path
- For each deployment descriptor, manager shall automatically create container and run container

Optional notes:
- It should be configurable whether an auto-created container it also automatically started (eg autorun:true)
- Watching the folder for changes of existing container deployment descriptors or new files is not necessary for a production environment and may be considered as an optional feature enabled at runtime for development purposes.

k3s has a similar feature called [Auto-Deploying Manifests](https://docs.k3s.io/advanced#auto-deploying-manifests).

A sample implementation in Rust exists at https://github.com/vasilvas99/talk-to-kanto

**Tasks**
- #103
- eclipse-kanto/kanto#191

**Comment by e-grigorov** [comment]
Is it a request for an initial only auto-deployment or ongoing one? I can see some differences. For example, when to remove the container?
- if it's an initial only auto-deployment, the answer should be: never
- if it's an ongoing auto-deployment, most probably it's expected to remove the container when its descriptor is removed

Why did you decide to do this as an external binary? Contribution to the container management is always an option :)

**Comment by mikehaller** [comment]
We started with a replacement of a bash utility to get the status of containers and since parsing the output kanto-cm wasn't so nice idea, we wanted to evaluate how to work with the proper container-management API and trying that out. At the same time, we were wondering how to do the auto deployment and it got mixed up.

In the meantime, it became clearer to us that the feature of auto-deployment of containers would be a much better fit as a feature in Kanto container-management instead of a separate application.

**Comment by mikehaller** [comment]
> Is it a request for an initial only auto-deployment or ongoing one? I can see some differences. For example, when to remove the container?
>     * if it's an initial only auto-deployment, the answer should be: never
>     * if it's an ongoing auto-deployment, most probably it's expected to remove the container when its descriptor is removed

TLDR: We only need initial auto-deployment.

From our use case point of view (automotive, with a separate OTA process), we're only looking for the initial deployment of containers. Right now, the suggested solution was to create shell scripts / systemd units to call "kanto-cm" cli tool during first boot. I dislike that idea as the kanto-cm cli tool's command line interface would become an official API and any changes to that may have unintended side effects or may break system behaviour on future updates.

Also, on a production device the kanto-cm cli tool would not even needed to be installed except for this initial deployment, which seems a waste of disk space. The same functionality implemented as a core feature in container-management would certainly be much less (maybe a few more kilobytes).

Regarding ongoing watching of the folder for changes to the deployment descriptors: During development and testing, it may be a nice add-on to have a file watcher. If this is actually considered, then I guess on delete or modification, the existing container would have to be removed and the new one (on change) be recreated. That would mean additional state to be persisted and managed. That may be additional effort and complexity which may not be suitable to have it implemented in the container-management for production devices and hence may not be in scope for Kanto.

There may exist other use cases as well, but they're currently not in my scope of needs (e.g. factory reset of device).

**Comment by e-grigorov** [comment]
@dimitar-dimitrow, I would suggest to start with a quick research/check about the possible manifest formats.

**Comment by dimitar-dimitrow** [comment]
We can either use format specific to kanto container management as the easiest and fastest solution or kubernetes manifest format as a more adopted one.

In fact there is already kanto container management specific [Manifest](https://github.com/eclipse/vorto/blob/development/models/com.bosch.iot.suite.edge.containers-Manifest-1.3.0.type/model.datatype) described in a vorto model, which is currently not used, but is up to date with the main branch. The Manifest could contains multiple [ContainerManifests](https://github.com/eclipse/vorto/blob/development/models/com.bosch.iot.suite.edge.containers-ContainerManifest-1.3.0.type/model.datatype) that represents containers(name + image reference + [Configuration](https://github.com/eclipse/vorto/blob/development/models/com.bosch.iot.suite.edge.containers-Configuration-1.3.0.type/model.datatype)).

In k8s manifest is API object description. A config file in JSON or YAML format that could contain one or more different resources. Such as a pod, replica set, secret, deployment and etc. Looking at the k8s documentation the closest to a kanto container is a k8s pod. Here are [examples](https://k8s-examples.container-solutions.com/examples/Pod/Pod.html) of k8s pod definitions.

It should be noted that there are significant differences between kanto container and k8s pod, the biggest one is that pod does not represent a single container, but a group of one or more containers, with shared storage and network resources. The container representation within a pod could be checked [here](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#Container). There are a some discrepancies when matched with a kanto container, e.g. additional ways to specify environment variables(config map, secrets, field ref, resource field ref), volumes instead of mount points, lifecycle is missing in kanto and etc. There are also some things that are in kanto container and missing in k8s - support for encrypted container images and verification of signed container images, which is currently in progress. Fortunately, decryption and signature verification configurations could also be specified globally on container management start and not only per container.

|   | **kanto format** | **k8s format** |
|---|-------|-----|
| **Adoption**  | Kanto specific.  |   Widely adopted. State of the art.  |
| **Mapping**  | Straightforward.  |  Confusing (only Pod support and its partial).  |
| **Binary size increase**  |   No significant changes are expected.    |   Would go up in size (+7mb only for k8s.io/api and +9mb for k8s.io/api and sigs.k8s.io/yaml)  |
| **Migration from k8s to kanto** |   The user should create its manifest from scratch. E.g. re-create its initial setup and export it. |  The user should get known with mapping limitations.  |
|  **Opportunities** | Stay open for new possibilities.  | Urge for further integration with k8s.  |
|  **Time to market** |   Less(should be further estimated).    |   More(should be further estimated).  |

**Comment by mikehaller** [comment]
+1 for kanto format

On embedded, you want to save as much space as possible.

Additional formats could be supported by thirdparty addons and do not need to be built into kanto imho.

**Comment by dimitar-dimitrow** [comment]
There are difficulties(cyclic dependencies) to handle the Manifest format outside from the things package, that were not taken in account during the research.

So I defaulted to the common [Kanto Container format](https://websites.eclipseprojects.io/kanto/docs/references/containers/container-config/#template). JSON files with such container descriptors provided in a daemon configurable directory would be loaded on initial startup.

**Comment by dimitar-dimitrow** [comment]
@mikehaller, container descriptors for initial deployment could now be provided in the deployment's init_dir configuration, which is described in the [documentation](https://websites.eclipseprojects.io/kanto/docs/references/containers/container-manager-config/), by default it is  _/etc/container-management/containers_.
